crates/sparrow-compiler/src/data_context.rs

```diff
@@ -303,6 +303,52 @@ pub struct GroupInfo {
     key_type: DataType,
 }
 
+/// Concurrent file sets allows us to update the file sets from the
+/// python library and see updates reflected within the data context.
+#[derive(Debug, Default, Clone)]
+pub struct ConcurrentFileSets {
+    file_sets: Arc<Mutex<Vec<compute_table::FileSet>>>,
+}
+
+impl ConcurrentFileSets {
+    pub fn new(file_sets: Vec<compute_table::FileSet>) -> Self {
+        Self {
+            file_sets: Arc::new(Mutex::new(file_sets)),
+        }
+    }
+
+    pub fn push(&mut self, file_set: compute_table::FileSet) {
+        self.file_sets.lock().unwrap().push(file_set);
+    }
+
+    pub fn append(&mut self, slice_plan: Option<v1alpha::SlicePlan>, prepared: Vec<PreparedFile>) {
+        let mut file_sets = self.file_sets.lock().unwrap();
+
+        let file_set = file_sets
+            .iter_mut()
+            .find(|set| set.slice_plan == slice_plan);
+
+        if let Some(file_set) = file_set {
+            file_set.prepared_files.extend(prepared);
+        } else {
+            file_sets.push(compute_table::FileSet {
+                slice_plan,
+                prepared_files: prepared,
+            });
+        }
+    }
+
+    // TODO: Cloning is bad, but since it's locked behind a mutex, it's necessary.
+    // Can we do better?
+    pub fn get_clone(&self) -> Vec<compute_table::FileSet> {
+        self.file_sets.lock().unwrap().clone()
```

Comment:
This should only be necessary if we need ownership of the file sets. We should be able to do something like:

```
pub fn file_sets(&self) -> MutexGuard<'_, Vec<compute_table::FileSet>> {
  self.file_sets.lock().unwrap()
}
```


Note `unwrap` here is a potential source of a panic if the mutex has been poisoned. Probably OK (and should just not let the lock get poisoned).

Reply:
Hmm. The problem is at least partly that the consumer wants to have the owned file sets and work with them (and probably for lifetime reasons, the guarded reference wouldn't work). I wonder if we'd be better off with a `Mutex<RefCell<Arc<[FileSet]>>>`. Specifically, when we add files, we could create a new Arc but we could then create a copy of the fileset rather than mutating it. This would allow us to just copy the current file set and compute on it.

Reply:
Hm..that still complains for me -- if the intention is to be able to do 
```
    pub fn get(&self) -> Ref<'_, Vec<compute_table::FileSet>> {
        self.file_sets.lock().unwrap().borrow()
    }
```

It's creating a temporary variable with the `lock().unwrap()`.

I can do:
```
    pub fn get(&mut self) -> Ref<'_, Vec<compute_table::FileSet>> {
        self.file_sets.get_mut().unwrap().borrow()
    }
```
but the `&mut self` is perhaps not ideal.

Reply:
We do want to be able to mutate the `FileSet` within the `Mutex`, which I believe you can't do if it's wrapped by an `Arc`?